common/src/main/java/com/microsoft/identity/common/internal/providers/oauth2/CameraPermissionRequestHandler.kt

```diff
@@ -97,19 +142,92 @@ class CameraPermissionRequestHandler {
+                && PreferredAuthMethod.QR.value.equals(SdmQrPinManager.getPreferredAuthConfig())
+    }
+
+    /**
+     * Handles the camera permission request for the app.
+     * If the camera permission is already granted, it will grant the permission.
+     * Otherwise, it will request the camera permission.
+     */
+    private fun defaultHandler(context: Context) {
+        val methodTag = "$TAG:defaultHandler"
+        if (appHasCameraPermission(context)) {
+            Logger.info(methodTag, "App level camera permission already granted, silent grant.")
+            grant()
+        } else {
+            requestCameraPermission()
+        }
+    }
+
+    /**
+     * Handles the camera permission request for QR code scanning.
+     * If the camera permission is already granted and the camera consent suppress is enabled,
+     * it will grant the permission. otherwise, it will explaining why the camera permission is required.
+     * If the camera permission is not granted, it will request the camera permission.
+     */
+    private fun qrPinHandler(context: Context) {
+        val  methodTag = "$TAG:handleQrPin"
+        if (appHasCameraPermission(context)) {
+            Logger.info(methodTag, "App level camera permission already granted.")
+            if (SdmQrPinManager.isCameraConsentSuppressed()) {
+                Logger.info(methodTag, "Camera consent suppress is enabled.")
+                grant()
+            } else {
+                Logger.info(methodTag, "Camera consent suppress is not enabled.")
+                showQrPinCameraRationale(context)
+            }
+        } else {
+            requestCameraPermission()
+        }
+    }
+
+    /**
+     * Determines whatever if the camera permission has been granted.
      *
-     * @param permissionRequest The permission request to set.
+     * @return true if the camera permission has been granted, false otherwise.
+     */
+    private fun appHasCameraPermission(context: Context): Boolean {
+        return (ContextCompat.checkSelfPermission(context, Manifest.permission.CAMERA)
+                == PackageManager.PERMISSION_GRANTED)
+    }
+
+    /**
+     * Shows a dialog to the user explaining why the camera permission is required.
+     * If the user accepts the dialog, the camera permission request will be launched.
+     * If the user denies the dialog, the camera permission request will be denied.
      */
-    private fun setCurrentPermissionRequest(permissionRequest: PermissionRequest) {
-        currentPermissionRequest = permissionRequest
-        isGranted = false
+    private fun showQrPinCameraRationale(context: Context) {
+        val methodTag = "$TAG:showQrPinCameraRationale"
+        val builder = AlertDialog.Builder(context)
+        builder.setMessage(R.string.qr_code_rationale_message)
+            .setTitle(R.string.qr_code_rationale_header)
+            .setCancelable(false)
```

Comment:
Why is cancelable set to false?

Reply:
In the showQrPinCameraRationale method, the setCancelable(false) is used to ensure that the dialog cannot be dismissed by tapping outside the dialog or pressing the back button. This is likely done to make sure the user explicitly chooses one of the provided options (allow or block) for granting or denying the camera permission.